### PR TITLE
Pass URDF to controllers on init

### PR DESCRIPTION
**Pull request by bmagyar**
This should help avoiding extra legwork in controllers to get access to the `/robot_description`.

### CODE REVIEW

File: controller_interface/include/controller_interface/controller_interface_base.hpp [hunk 1]
```diff
@@ -223,6 +223,7 @@ class ControllerInterfaceBase : public rclcpp_lifecycle::node_interfaces::Lifecy
   std::vector<hardware_interface::LoanedStateInterface> state_interfaces_;
   unsigned int update_rate_ = 0;
   bool is_async_ = false;
+  std::string urdf_ = "";
```
Comment (aprotyas) [inline controller_interface/include/controller_interface/controller_interface_base.hpp:226]: Should this be a `std::optional<std::string> urdf_ = std::nullopt` instead?

**Review by aprotyas — APPROVED** [review]

**Review by saikishor — COMMENTED** [review]
The provision of the URDF directly to the controllers is a brilliant idea. This reduces the hassles one should go through to set up their controllers. 
https://github.com/ros-controls/ros2_control/blob/757ccd45b7a70d226e902e2dcbba0ff96f31ace2/controller_manager/src/controller_manager.cpp#L1143

However, I think there is a missing part in assigning this URDF from the controller_manager's end, and at least one test that verifies the robot_description string from the topic or param is the same as that inside the controller. 
Another test that checks when a URDF is updated in the topic, the newly initialized controllers get the new URDF rather than the old one.

Brilliant work :clap:

**Review by destogl — DISMISSED** [review]
This looks good!

File: controller_interface/include/controller_interface/controller_interface_base.hpp (hunk 1 above)
Comment (bmagyar) [reply controller_interface/include/controller_interface/controller_interface_base.hpp:227]: I don't want to communicate the option of not being set as anything close to valid. The empty string initial value is only there so we can test if it was properly set...

I also thought about how we could do this better, a const string would be nice but that can only be set once which doesn't work well with the longer term idea of getting URDF updates from topics.

File: controller_manager/test/test_controller_manager.cpp [hunk 2]
```diff
@@ -33,6 +33,34 @@ class TestControllerManagerWithStrictness
...
+class TestControllerManagerRobotDescription
+: public ControllerManagerFixture<controller_manager::ControllerManager>
+{
+};
+
+TEST_F(TestControllerManagerRobotDescription, controller_robot_description_update)
+{
+  auto test_controller = std::make_shared<test_controller::TestController>();
+  auto test_controller2 = std::make_shared<test_controller::TestController>();
+  constexpr char TEST_CONTROLLER2_NAME[] = "test_controller2_name";
```
Comment (bmagyar) [inline controller_manager/test/test_controller_manager.cpp:45]: Could this be auto without []? And maybe outside of this function

File: controller_manager/test/test_controller_manager.cpp [hunk 3]
```diff
@@ -33,6 +33,34 @@ class TestControllerManagerWithStrictness
...
+  ASSERT_EQ(ros2_control_test_assets::minimal_robot_urdf, test_controller->getRobotDescription());
+
+  // Now change the robot description and then load a new controller and see if the new controller
+  // gets the new description and the old controller still maintains the configuration
+  auto msg = std_msgs::msg::String();
+  msg.data = ros2_control_test_assets::minimal_robot_missing_state_keys_urdf;
+  cm_->robot_description_callback(msg);
+  cm_->add_controller(
+    test_controller2, TEST_CONTROLLER2_NAME, test_controller::TEST_CONTROLLER_CLASS_NAME);
+  ASSERT_EQ(ros2_control_test_assets::minimal_robot_urdf, test_controller->getRobotDescription());
```
Comment (bmagyar) [inline controller_manager/test/test_controller_manager.cpp:58]:
```suggestion
  ASSERT_EQ(ros2_control_test_assets::minimal_robot_urdf, test_controller2->getRobotDescription());
```

And you'll need another one if (as per the comment) you want to check the old controller having access to the old urdf

File: controller_manager/test/test_controller_manager_srvs.cpp [hunk 4]
```diff
@@ -456,6 +456,47 @@ TEST_F(TestControllerManagerSrvs, unload_controller_srv)
...
+
+  // check the robot description
+  ASSERT_EQ(ros2_control_test_assets::minimal_robot_urdf, test_controller->getRobotDescription());
+
+  // Now change the robot description and then load a new controller and see if the new controller
+  // gets the new description and the old controller still maintains the configuration
+  auto msg = std_msgs::msg::String();
+  msg.data = ros2_control_test_assets::minimal_robot_missing_state_keys_urdf;
+  cm_->robot_description_callback(msg);
+  ASSERT_EQ(ros2_control_test_assets::minimal_robot_urdf, test_controller->getRobotDescription());
```
Comment (bmagyar) [inline controller_manager/test/test_controller_manager_srvs.cpp:482]: How about controller2?

File: controller_manager/test/test_controller_manager.cpp (hunk 2 above)
Comment (saikishor) [reply controller_manager/test/test_controller_manager.cpp]: Sure. Done!

File: controller_manager/test/test_controller_manager_srvs.cpp (hunk 4 above)
Comment (saikishor) [reply controller_manager/test/test_controller_manager_srvs.cpp:482]: Sorry,  the above comments are misleading. this needs to be changes. I have updated the comments in the recent commit. Thanks for pointing out.

Comment (bmagyar) [reply controller_manager/test/test_controller_manager_srvs.cpp]: thank you!

File: controller_manager/test/test_controller_manager.cpp (hunk 3 above)
Comment (bmagyar) [reply controller_manager/test/test_controller_manager.cpp]: thx

**Comment by sea-bass** [comment]
Hey @bmagyar -- I am working with some folks that are using Humble and would like this feature backported. Would you be okay with this, or does this break ABI in some way?

EDIT: Yeah this definitely breaks ABI... ugh...

**Comment by christophfroehlich** [comment]
Yes, I don't think it makes sense to backport this.
Are you aware that we support building the current development version (aka rolling) on humble? 5min build time in your workspace, and you've got all the new features ;)

**Comment by sea-bass** [comment]
I was not aware it still built on humble, thanks!

**Comment by bmagyar** [comment]
I'm afraid it's break too many things for backport it.

**Comment by saikishor** [comment]
> I was not aware it still built on humble, thanks!

@sea-bass You can find more information here:
https://control.ros.org/humble/doc/getting_started/getting_started.html#building-from-source

**Pull request merged**